### PR TITLE
Fix 404 contact not found with shared addressbook

### PR DESCRIPTION
**Pull request by LEDfan**
There are again 404 occuring, this time with a shared addressbook. The contacts app tries to search for a contact in a `local` addressbook, but it fails because the addressbook isn't owned by this user. Instead it should access the same addressbook but not by using the `local` backend but the `shared` backend.

Refs https://github.com/owncloud/chat/issues/159#issuecomment-63945487

Note that this solution is a hack! In the DB there isn't a field which backend is used for this contact, thus we should guess this. Luckily the contact id is unique -> there can't be a match in the local addressbook when the contact is stored in the shared addressbook.

### CODE REVIEW

**Comment by LEDfan** [comment]
cc @babelouest @jbtbnl @DeepDiver1975

File: lib/addressbookprovider.php [hunk 1]
```diff
@@ -266,4 +271,19 @@ private function convertToSearchResult($j) {
 
 		return $result;
 	}
+
+	private function getAddresbookKeyForContact($id){
+		$table = self::CONTACT_TABLE;
+		$query = <<<SQL
```
Comment (LEDfan) [inline lib/addressbookprovider.php]: I'll add this to the main query.

Comment (LEDfan) [inline lib/addressbookprovider.php]: Done

File: lib/addressbookprovider.php [hunk 2]
```diff
@@ -187,6 +206,7 @@ public function createOrUpdate($properties) {
 					break;
 				case 'EMAIL':
 				case 'TEL':
+				case 'TEL':
```
Comment (jbtbnl) [inline lib/addressbookprovider.php]: I think this is a typo :)

Comment (LEDfan) [inline lib/addressbookprovider.php]: Indeed, thanks!

**Comment by Gomez** [comment]
:+1: Works now! Get it in stable7 isn't a problem, right?

File: lib/addressbookprovider.php [hunk 3]
```diff
@@ -89,11 +90,26 @@ public function getPermissions() {
...
+				`$propTable`.`contactid`,
+				`$contTable`.`addressbookid`
+			FROM
+				`$propTable`
+			INNER JOIN
+				`$contTable`
+			ON `$contTable`.`id` = `$propTable`.`contactid`
+			WHERE
+			  `$propTable`.`userid` = ?
+			AND (
```
Comment (jbtbnl) [inline lib/addressbookprovider.php]: Whats up with this AND?

**Comment by LEDfan** [comment]
@Gomez :fireworks: :fireworks: Nice! I'll backport it for 7.0.4.

**Comment by jbtbnl** [comment]
@LEDfan let's wait for a go from @babelouest before backporting :)

File: lib/addressbookprovider.php (hunk 3 above)
Comment (LEDfan) [inline lib/addressbookprovider.php]: It's used by lines 112-116, to add the search parameters.

Comment (jbtbnl) [inline lib/addressbookprovider.php]: Ah I see, thanks

**Comment by LEDfan** [comment]
Please don't merge, since there is still something wrong with this solution.

**Comment by LEDfan** [comment]
I added another commit so the shared contacts are returned too.

@Gomez can you test if it still works? Tnx

**Comment by Gomez** [comment]
@LEDfan Works fine. And i think this fixed the contacts count on shared addressbooks, too.

**Comment by babelouest** [comment]
I haven't test it yet, but the code looks clean.
I will test it with the ldap backend too, but I'm not even sure if the search engine works with this backend :p

**Comment by Gomez** [comment]
Can i do something further to get it merged?

**Comment by Gomez** [comment]
Still possible to get this in 7.0.4? Would be good!

**Comment by LEDfan** [comment]
@babelouest @jbtbnl can this be merged and backported to 7.0.4?

**Comment by babelouest** [comment]
I have problems using owncloud today: #724 , I will try to test it as soon as I can

**Comment by babelouest** [comment]
That works for me, a shared contact is searchable, no 404, although search doesn't work with ldap backend, this can be fix later.

**Comment by Gomez** [comment]
Oh, can we backport into stable7

**Comment by Gomez** [comment]
@LEDfan Can we backport to stable7? I would use this branch on my installations, to bad it hasnt got into 7.0.4 :(

**Pull request merged**